File: internal/tools/grep.go
package tools

import (
	"context"
	"os/exec"
	"strings"

	"minicode/internal/domain"
)

// GrepTool wraps the system grep command.
type GrepTool struct{}

// NewGrepTool creates a GrepTool.
func NewGrepTool() *GrepTool { return &GrepTool{} }

func (t *GrepTool) Name() string { return "Grep" }
func (t *GrepTool) Description() string {
	return "Search file contents using the system grep command. Supports recursive search, line numbers, and case-insensitive matching. Searches are restricted to the current working directory and its subdirectories."
}
func (t *GrepTool) ReadOnly() bool    { return true }
func (t *GrepTool) Interactive() bool { return false }
func (t *GrepTool) Requires() []ToolRequirement { return nil }

func (t *GrepTool) InputSchema() map[string]any {
	return map[string]any{
		"type": "object",
		"properties": map[string]any{
			"pattern":     map[string]any{"type": "string", "description": "The regex pattern to search for"},
			"path":        map[string]any{"type": "string", "description": "Directory or file to search in (default: current directory)"},
			"recursive":   map[string]any{"type": "boolean", "description": "Search recursively in subdirectories (default: true)"},
			"ignore_case": map[string]any{"type": "boolean", "description": "Case-insensitive search"},
			"include":     map[string]any{"type": "string", "description": "File glob pattern to include, e.g. '*.go'"},
		},
		"required": []string{"pattern"},
	}
}

func (t *GrepTool) Execute(ctx context.Context, args map[string]any, tc ToolContext) (domain.ToolResult, error) {
	pattern, _ := args["pattern"].(string)
	if pattern == "" {
		return domain.ToolResult{Output: "Error: pattern is required"}, nil
	}

	path, _ := args["path"].(string)
	if path == "" {
		path = "."
	}

	recursive := true
	if r, ok := args["recursive"].(bool); ok {
		recursive = r
	}

	ignoreCase := false
	if i, ok := args["ignore_case"].(bool); ok {
		ignoreCase = i
	}

	include, _ := args["include"].(string)

	grepArgs := []string{"-n"}
	if ignoreCase {
		grepArgs = append(grepArgs, "-i")
	}
	if recursive {
		grepArgs = append(grepArgs, "-r")
	}
	if include != "" {
		grepArgs = append(grepArgs, "--include="+include)
	}
	grepArgs = append(grepArgs, pattern, path)

	tc.Log("Grep: executing", "pattern", pattern, "path", path, "recursive", recursive, "ignore_case", ignoreCase)

	cmd := exec.CommandContext(ctx, "grep", grepArgs...)
	output, err := cmd.CombinedOutput()

	outputStr := strings.TrimSpace(string(output))
	if err != nil {
		if ctx.Err() == context.Canceled {
			return domain.ToolResult{Output: "Aborted"}, nil
		}
		if outputStr == "" {
			return domain.ToolResult{Output: "No matches found"}, nil
		}
		return domain.ToolResult{Output: outputStr}, nil
	}

	tc.Log("Grep: completed", "matches", strings.Count(outputStr, "\n")+1)
	return domain.ToolResult{Output: outputStr}, nil
}

func init() { Register(NewGrepTool()) }

File: src/tools/grep.ts
import { spawn } from "child_process";
import type { ToolDef, ToolResult, ToolExecutionContext } from "./registry.js";
import { register } from "./registry.js";

export const grepTool: ToolDef = {
  name: "Grep",
  description:
    "Search file contents using the system grep command. Supports recursive search, line numbers, and case-insensitive matching. Searches are restricted to the current working directory and its subdirectories.",
  requiresPermission: false,
  readOnly: true,
  input_schema: {
    type: "object" as const,
    properties: {
      pattern: {
        type: "string",
        description: "The regex pattern to search for",
      },
      path: {
        type: "string",
        description:
          "Directory or file to search in (default: current directory)",
      },
      recursive: {
        type: "boolean",
        description: "Search recursively in subdirectories (default: true)",
      },
      ignore_case: { type: "boolean", description: "Case-insensitive search" },
      include: {
        type: "string",
        description: "File glob pattern to include, e.g. '*.ts'",
      },
    },
    required: ["pattern"],
  },
  execute: async (
    args: Record<string, unknown>,
    context?: ToolExecutionContext,
  ): Promise<ToolResult> => {
    const pattern = args.pattern as string;
    if (!pattern) return { output: "Error: pattern is required" };

    const searchPath = (args.path as string) || ".";
    const recursive = args.recursive !== false;
    const ignoreCase = !!args.ignore_case;
    const include = args.include as string | undefined;

    const grepArgs = ["-n"];
    if (ignoreCase) grepArgs.push("-i");
    if (recursive) grepArgs.push("-r");
    if (include) grepArgs.push(`--include=${include}`);
    grepArgs.push(pattern, searchPath);

    try {
      const output = await new Promise<string>((resolve) => {
        const proc = spawn("grep", grepArgs, {
          stdio: ["pipe", "pipe", "pipe"],
        });

        let stdout = "";
        let stderr = "";
        proc.stdout?.on("data", (d) => (stdout += d.toString()));
        proc.stderr?.on("data", (d) => (stderr += d.toString()));

        context?.signal?.addEventListener("abort", () => proc.kill());

        proc.on("close", () => {
          const trimmed = (stdout || stderr).trim();
          resolve(trimmed || "No matches found");
        });
      });
      return { output };
    } catch (err) {
      const msg = err instanceof Error ? err.message : String(err);
      return { output: msg };
    }
  },
};
register(grepTool);
